accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -495,10 +502,24 @@ of a function call produces a result of type `Never`.
 
 The static type of a `throw e` expression is `Never`.
 
-In legacy mode, an override of operator== with no explicit parameter type
+Consider an expression `e` of the form `e1 == e2` where the static type of
+`e1` is `T1` and the static type of `e2` is `T2`. Let `S` be the type of the
+formal parameter of `operator ==` in the interface of **NonNull**(`T1`).
+It is a compile-time error unless `T2` is assignable to `S?`.
```

Comment:
I think this is a mistake, and we should not do it.

No-one should create an operator == which takes anything except `Object`. A proper implementation must gracefully accept any object as argument and *must not throw*.

We should not support giving the operator method a parameter other than `Object`, and we should not even suggest that it's in any way acceptable.

I'd say that:

> Consider and expression `e` of the form `e1 == e2`. The static type of `e` is `bool`.
> It is a compile-time error if the static type of `e1` or `e2` is `void` or `void*`.

That's all we need.

The semantics should not be worrying about what type `e1` actually has, it should just cast it and `e2` to `Object`, then call `operator==`  on that. If that throws, someone made a mistake and they should fix it.

I would support a hint that warned you if you use anything but `Object` as the parameter type of `operator==`.

Reply:
I don't care strongly about this one way or the other.  I think though that you are proposing more than just undoing this change: currently, you can covariantly override the parameter to == on class A and have that respected.  This change is just around whether == on `A?` is treated the same as == on `A`, or differently.  While I don't care much, I don't see a great argument for respecting it for == on A and not on `A?`.  I think you're advocating for never respecting it.  I'm amenable to discussing, maybe open a language issue on it?

Reply:
The current spec says only:

> The static type of an equality expression is \code{bool}.

There is no concern for the declaration of `operator==` on the first operand. That's (among other things) what this change is changing, and what I want undone.

If the current implementations are doing something different, they are wrong. That seems to apply to the analyzer, the CFE does not complain (it just throws at run-time like it should for any mal-called covariant parameter).

That is, you *cannot* have your covariant override respected in general, only in the analyzer, and that is a bug according to the specification.

I'm opposing changing the specification to match the analyzer. I'd prefer to keep it matching the CFE.

I'll open an issue for discussing this.

Reply:
> I'm opposing changing the specification to match the analyzer. I'd prefer to keep it matching the CFE.

I don't understand where you're getting this.  The following code is rejected by both the analyzer and the CFE:

```dart
class A {
  operator ==(covariant int x) => false;
}

void main() {
  var x = new A();
  x == "hello";
}
```

Reply:
My impression of CFE comes from checking in DartPad and locally on a 2.7.2 VM. If the CFE has changed, it is a recent change.

Reply:
I've been convinced that this isn't that bad, and we are already doing it. (Which kindof proves that it isn't bad).